src/libprojectM/PCM.cpp

```diff
@@ -24,381 +24,528 @@
+    // per rdft() documentation
+    //    length of ip >= 2+sqrt(n) and length of w == n/2
 #if FFT_LENGTH > 1024
 #error update this code
 #endif
-  w  = (double *)wipemalloc(FFT_LENGTH/2*sizeof(double));
-  ip = (int *)wipemalloc(34 * sizeof(int));
-  ip[0]=0;
-
-    /** PCM data */
-//    this->maxsamples = 2048;
-//    this->numsamples = 0;
-//    this->pcmdataL = NULL;
-//    this->pcmdataR = NULL;
+    w  = (double *)wipemalloc(FFT_LENGTH*sizeof(double));
+    ip = (int *)wipemalloc(34 * sizeof(int));
+    ip[0]=0;
+
+    memset(pcmL, 0, sizeof(pcmL));
+    memset(pcmR, 0, sizeof(pcmR));
+    memset(freqL, 0, sizeof(freqL));
+    memset(freqR, 0, sizeof(freqR));
+    memset(spectrumL, 0, sizeof(spectrumL));
+    memset(spectrumR, 0, sizeof(spectrumR));
+}
 
-    /** Allocate PCM data structures */
-    pcmdataL=(float *)wipemalloc(this->maxsamples*sizeof(float));
-    pcmdataR=(float *)wipemalloc(this->maxsamples*sizeof(float));
 
+PCM::~PCM()
+{
+    delete leveler;
+    free(w);
+    free(ip);
 }
 
-PCM::~PCM() {
-
-	free(pcmdataL);
-	free(pcmdataR);
-	free(w);
-	free(ip);
+#include <iostream>
 
-	free(PCMd[0]);
-	free(PCMd[1]);
-	free(PCMd);
 
+void PCM::addPCMfloat(const float *PCMdata, size_t samples)
+{
+    float a,sum=0,max=0;
+    for (size_t i=0; i<samples; i++)
+    {
+        size_t j=(i+start)%maxsamples;
+        a=pcmL[j] = PCMdata[i];
+        pcmR[j] = PCMdata[i];
+        sum += fabs(a);
+        max = fmax(max,a);
+    }
+    start = (start+samples)%maxsamples;
+    newsamples += samples;
+    level = leveler->updateLevel(samples, sum, max);
 }
 
-#include <iostream>
 
-void PCM::addPCMfloat(const float *PCMdata, int samples)
+/* NOTE: this method expects total samples, not samples per channel */
+void PCM::addPCMfloat_2ch(const float *PCMdata, size_t count)
 {
-  int i,j;
+    size_t samples = count/2;
+    float a,b,sum=0,max=0;
+    for (size_t i=0; i<samples; i++)
+    {
+        size_t j=(start+i)%maxsamples;
+        a = pcmL[j] = PCMdata[i*2];
+        b = pcmR[j] = PCMdata[i*2+1];
+        sum += fabs(a) + fabs(b);
+        max = fmax(fmax(max,fabs(a)),fabs(b));
+    }
+    start = (start + samples) % maxsamples;
+    newsamples += samples;
+    level = leveler->updateLevel(samples, sum/2, max);
+}
+
 
-  for(i=0;i<samples;i++)
+void PCM::addPCM16Data(const short* pcm_data, size_t samples)
+{
+    float a, b, sum = 0, max = 0;
+    for (size_t i = 0; i < samples; ++i)
     {
-      j=i+start;
+        size_t j = (i + start) % maxsamples;
+        a = pcmL[j] = (pcm_data[i * 2 + 0] / 16384.0);
+        b = pcmR[j] = (pcm_data[i * 2 + 1] / 16384.0);
+        sum += fabs(a) + fabs(b);
+        max = fmax(fmax(max, a), b);
+    }
+    start = (start + samples) % maxsamples;
+    newsamples += samples;
+    level = leveler->updateLevel(samples, sum/2, max);
+}
 
-      if (PCMdata[i] != 0 ) {
 
-	PCMd[0][j%maxsamples] = PCMdata[i];
-	PCMd[1][j%maxsamples] = PCMdata[i];
+void PCM::addPCM16(const short PCMdata[2][512])
```

Comment:
Are we assuming there is always 2048 bytes of samples? Can there be a constant for this?
I want to experiment with fewer samples, maybe it can make the visuals more responsive?